lxd/api_cluster.go

```diff
@@ -4599,3 +4618,675 @@ func autoHealCluster(_ context.Context, s *state.State, offlineMembers []db.Node
+	}
+
+	// Get the existing cluster link.
+	err = s.DB.Cluster.Transaction(context.TODO(), func(ctx context.Context, tx *db.ClusterTx) error {
+		_, err := dbCluster.GetClusterLink(ctx, tx.Tx(), name)
+		if err != nil {
+			return fmt.Errorf("failed to fetch cluster link: %w", err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	// Update DB entry.
+	err = s.DB.Cluster.Transaction(context.TODO(), func(ctx context.Context, tx *db.ClusterTx) error {
+		err = dbCluster.RenameClusterLink(ctx, tx.Tx(), name, req.Name)
+		if err != nil {
+			return err
+		}
+
+		return err
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	// Send cluster link lifecycle event.
+	requestor := request.CreateRequestor(r)
+	s.Events.SendLifecycle(api.ProjectDefaultName, lifecycle.ClusterLinkUpdated.Event(name, requestor, nil))
+
+	return response.SyncResponse(true, name)
+}
+
+// swagger:operation DELETE /1.0/clusters/{name} cluster-links cluster_link_delete
+//
+//	Delete cluster link
+//
+//	Delete a cluster link.
+//
+//	---
+//	produces:
+//	  - application/json
+//	responses:
+//	  "200":
+//	    description: Delete cluster link request
+//	    schema:
+//	      type: object
+//	      description: Sync response
+//	      properties:
+//	        type:
+//	          type: string
+//	          description: Response type
+//	          example: sync
+//	        status:
+//	          type: string
+//	          description: Status description
+//	          example: Success
+//	        status_code:
+//	          type: integer
+//	          description: Status code
+//	          example: 200
+//	  "403":
+//	    $ref: "#/responses/Forbidden"
+//	  "500":
+//	    $ref: "#/responses/InternalServerError"
+func clusterLinkDelete(d *Daemon, r *http.Request) response.Response {
+	s := d.State()
+
+	name, err := url.PathUnescape(mux.Vars(r)["name"])
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	// If a request tries to create/update/delete a delegated cluster link, the correct user agent must be set to prevent accidental edits of delegated cluster links.
+	if isDelegatedClusterLink(r) && r.Header.Get("User-Agent") != clusterRequest.UserAgentDelegator {
+		return response.Forbidden(fmt.Errorf("cannot update delegated cluster link %q", name))
+	}
+
+	// Update DB entry.
+	err = s.DB.Cluster.Transaction(context.TODO(), func(ctx context.Context, tx *db.ClusterTx) error {
+		err := dbCluster.DeleteClusterLink(ctx, tx.Tx(), name)
+		if err != nil {
+			return err
+		}
+
+		err = dbCluster.DeleteClusterLinkIdentity(ctx, tx.Tx(), name)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(fmt.Errorf("Error deleting %q from database: %w", name, err))
+	}
+
+	// Send cluster link lifecycle event.
+	requestor := request.CreateRequestor(r)
+	s.Events.SendLifecycle(api.ProjectDefaultName, lifecycle.ClusterLinkRemoved.Event(name, requestor, nil))
```

Comment:
Maybe two lifecycle events? An identity is being removed also. Discuss with Tom and Julian as well.

Reply:
yes please, same for cluster link add (when the pending identity is added).

And also emit the same lifecycle events for when the identity is activated (if that is what occurs for pending TLS identities).